File: 2 período/aeds2/beecrowd/q1766.c
// https://judge.beecrowd.com/pt/problems/view/1766

#include <stdio.h>

struct rena {
  char nome[100];
  int peso;
  int idade;
  float altura;
};


void ordenar(struct rena* vet, int n) {
    for (int i = 1; i < n; i++) {
        struct rena key = vet[i];
        int j = i - 1;

        // Ordena de acordo com os critérios especificados
        while (j >= 0 && 
               (vet[j].peso < key.peso ||  // Peso decrescente
               (vet[j].peso == key.peso && vet[j].idade > key.idade) ||  // Idade crescente
               (vet[j].peso == key.peso && vet[j].idade == key.idade && vet[j].altura > key.altura) ||  // Altura crescente
               (vet[j].peso == key.peso && vet[j].idade == key.idade && vet[j].altura == key.altura && strcmp(vet[j].nome, key.nome) > 0))) {  // Nome crescente
            vet[j + 1] = vet[j];  // Move o elemento para a direita
            j--;
        }
        vet[j + 1] = key;  // Insere o elemento na posição correta
    }
}

int main() {
  int t; scanf("%i", &t);

  while (t--) {
    int n; scanf("%i", &n);
    int m; scanf("%i", &m);

    struct rena vet[n];

    for (int i = 0; i < n; i ++) {
      scanf("%s %i %i %f", vet[i].nome, &vet[i].peso, &vet[i].idade, &vet[i].altura);
    }

    ordenar(vet, n);

    for (int i = 0; i < m; i ++) {
      printf("%s\n", vet[i].nome);
    }
  }

  return 0;
}

File: 2 período/aeds2/beecrowd/q1766.java
// https://judge.beecrowd.com/pt/problems/view/1766

import java.util.*;

class Renas {
    String nome;
    int peso;
    int idade;
    float altura;

    Renas(String nome, int peso, int idade, float altura) {
        this.nome = nome;
        this.peso = peso;
        this.idade = idade;
        this.altura = altura;
    }
}

public class q1766 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int t = sc.nextInt(); // lê o número de casos de teste
        sc.nextLine(); // consome a nova linha após ler t

        for (int i = 0; i < t; i++) {
            int n = sc.nextInt(); // lê n
            int m = sc.nextInt(); // lê m
            sc.nextLine(); // consome a nova linha após ler m

            Renas[] vet = new Renas[n];
            // Leitura dos dados dos participantes
            for (int j = 0; j < n; j++) {
                String input = sc.nextLine();
                String[] separadas = input.split(" ");
                int peso = Integer.parseInt(separadas[1]);
                int idade = Integer.parseInt(separadas[2]);
                float altura = Float.parseFloat(separadas[3]);

                vet[j] = new Renas(separadas[0], peso, idade, altura);

            }

            ordena(vet);

            for (int j = 0; j < m; j ++) {
                System.out.println(vet[j].nome);
            }

        }

        sc.close(); // fecha o Scanner
    }

    static void ordena(Renas vet[]) {
        for (int i = 1; i < vet.length; i++) {
            Renas key = vet[i];
            int actual = i;
    
            // O loop continua enquanto o índice atual for maior que 0
            // e as condições de ordenação forem satisfeitas.
            while (actual > 0 && 
                   (vet[actual - 1].peso < key.peso ||  // Peso decrescente
                   (vet[actual - 1].peso == key.peso && vet[actual - 1].idade > key.idade) ||  // Idade crescente
                   (vet[actual - 1].peso == key.peso && vet[actual - 1].idade == key.idade && vet[actual - 1].altura > key.altura) ||  // Altura crescente
                   (vet[actual - 1].peso == key.peso && vet[actual - 1].idade == key.idade && vet[actual - 1].altura == key.altura && vet[actual - 1].nome.compareTo(key.nome) > 0))) {  // Nome crescente
                vet[actual] = vet[actual - 1];
                actual--;
            }
    
            vet[actual] = key;
        }
    }
    
    
}
